### PR TITLE
feat: batch of teacher statistics

### PR DESCRIPTION
**Pull request by wielopolski**
## Jira issue(s)
[418](https://selleolabs.atlassian.net/jira/software/projects/LC/boards/3?selectedIssue=LC-418)
[419](https://selleolabs.atlassian.net/jira/software/projects/LC/boards/3?selectedIssue=LC-419)
[420](https://selleolabs.atlassian.net/jira/software/projects/LC/boards/3?selectedIssue=LC-420)

## Overview
This PR adds 3 metrics for teacher statistics:
- add conversion after freemium
- enrollments statistics
- average quiz score statistics

### CODE REVIEW

**Review by piotr-pajak — APPROVED** [review]

File: apps/api/src/questions/questions.service.ts [hunk 1]
```diff
@@ -94,11 +97,24 @@ export class QuestionsService {
         !quizProgress?.completedAt &&
         studentLessonProgress?.completedLessonItemCount === lesson.itemsCount
       ) {
+        const isCompletedAsFreemium = lesson.isFree && !lesson.enrolled;
```
Comment (typeWolffo) [inline apps/api/src/questions/questions.service.ts:100]: nitpick, but I think the name of that variable is a little bit confusing 🤔

File: apps/api/src/statistics/repositories/statistics.repository.ts [hunk 2]
```diff
@@ -137,23 +140,59 @@ export class StatisticsRepository {
...
-        totalCoursesCompletion: sql<number>`sum(courseStatisticsPerTeacherMaterialized.completedStudentCount)`,
-        totalCourses: sql<number>`sum(courseStatisticsPerTeacherMaterialized.studentCount)`,
+        totalCoursesCompletion: sql<number>`COALESCE(SUM(${coursesSummaryStats.completedCourseStudentCount}), 0)::INTEGER`,
+        totalCourses: sql<number>`COALESCE(SUM(${coursesSummaryStats.freePurchasedCount} + ${coursesSummaryStats.paidPurchasedCount}), 0)::INTEGER`,
+      })
+      .from(coursesSummaryStats)
+      .where(eq(coursesSummaryStats.authorId, userId));
+  }
+
+  async getConversationAfterFreemiumLesson(userId: string) {
```
Comment (typeWolffo) [inline apps/api/src/statistics/repositories/statistics.repository.ts:164]:
```suggestion
  async getConversionAfterFreemiumLesson(userId: string) {
```
probably copilot issue 😉

File: apps/api/src/statistics/repositories/statistics.repository.ts [hunk 3]
```diff
@@ -181,4 +220,74 @@ export class StatisticsRepository {
...
+    return dbInstance
+      .update(coursesSummaryStats)
+      .set({
+        completedFreemiumStudentCount: sql<number>`coursesSummaryStats.completedFreemiumStudentCount + 1`,
+      })
+      .where(eq(coursesSummaryStats.courseId, courseId));
+  }
+
+  async calculateCoursesStudentsStats(currentMonth: number, currentYear: number) {
+    const startOfMonth = new Date(currentYear, currentMonth, 1).toISOString();
```
Comment (typeWolffo) [inline apps/api/src/statistics/repositories/statistics.repository.ts:276]: we have date-fns which can be more convenient is such cases

File: apps/api/src/stripe/stripeWebhook.handler.ts [hunk 4]
```diff
@@ -78,6 +84,20 @@ export class StripeWebhookHandler {
           })),
         );
       }
+      const existingLessonProgress = await this.lessonsRepository.getLessonsProgressByCourseId(
+        course.id,
+        trx,
+      );
+
+      switch (existingLessonProgress.length) {
```
Comment (typeWolffo) [inline apps/api/src/stripe/stripeWebhook.handler.ts:92]: what do you think about ts-pattern with `isEmpty(existingLessonProgress)`?

**Review by typeWolffo — APPROVED** [review]
🟢 but I left some comments that can be resolved in future PRs

**Pull request merged**